Exception Tracking (#142)
rootstrap/blog: content/Exception Tracking - Tiziana Romani.md

# Exception Tracking

> ###### Production exceptions are crucial. We don't want errors encountered by clients or crashes in our application to go unnoticed, so keeping track of them it's a must! 

Reviewer: This doesn't need to be a heading

> ###### Production exceptions are crucial. We don't want errors encountered by clients or crashes in our application to go unnoticed, so keeping track of them it's a must! 

Reviewer 2: `keeping track of them is a must! ` instead of `it's`

[![Tulips](https://get.wallhere.com/photo/1920x1200-px-bokeh-fields-flowers-sea-tulips-white-1744269.jpg)](https://wallhere.com/es/wallpaper/1744269)

> An increase in software applications needs is appreciable nowadays. People want software to make life easier, thus the user experience must be rated as FFF: friendly, fast, and fine. Easy right? Not.

Reviewer: ```suggestion
An increase in software applications needs is appreciable nowadays. People want software to make life easier, thus the user experience must be rated as FFF: friendly, fast, and fine. Easy right? No.
```

> Every developer wishes their software to work like a charm, but since developers are human (at least for now), it is known that humans make mistakes. 

Reviewer: ```suggestion
Every developer wishes their software to work like a charm, but since developers are human (at least for now), it is known for them to make mistakes. 
```

And mistakes, normally sabotage the user experience... But raise up, what doesn't kill you makes you stronger. 

> Best way to go is by keeping this mistakes at bay. By monitoring them, awareness is key. But keeping record of hundreds maybe thousands (hopefully not) of errors, doesn't sound like an easy nor pleasant task. Yet diving through endless lines of logs isn't something you wish to someone, not even your worst enemy. There are easier ways, plenty of tools are awaiting for you to include them in your projects. Exception management tools. 

Reviewer: ```suggestion
The best way to go is by keeping these mistakes at bay. By monitoring them, awareness is key. But keeping record of hundreds maybe thousands (hopefully not) of errors, doesn't sound like an easy nor pleasant task. Yet diving through endless lines of logs isn't something you wish to someone, not even your worst enemy. There are easier ways, plenty of tools are waiting for you to include them in your projects. Exception management tools. 
```

> Best way to go is by keeping this mistakes at bay. By monitoring them, awareness is key. But keeping record of hundreds maybe thousands (hopefully not) of errors, doesn't sound like an easy nor pleasant task. Yet diving through endless lines of logs isn't something you wish to someone, not even your worst enemy. There are easier ways, plenty of tools are awaiting for you to include them in your projects. Exception management tools. 

Reviewer: ```suggestion
Best way to go is by keeping this mistakes at bay. By monitoring them, awareness is key. But keeping record of hundreds maybe thousands (hopefully not) of errors, doesn't sound like an easy nor pleasant task. Yet diving through endless lines of logs isn't something you wish to someone, not even your worst enemy. There are easier ways, plenty of tools exception management tools are waiting for you to include them in your projects. 
```

> Honeybadger, Sentry, Rollbar, are just a few names that resonate in this exception monitoring world. And don't take me wrong, they are great! But somehow all of them involve certain budget.  

Reviewer: ```suggestion
Honeybadger, Sentry, Rollbar, are just a few names that resonate in this exception monitoring world. And don't take me wrong, they are great! But somehow all of them involve a certain budget.  
```

 >“I think we can build a rocket ourselves.” 
 
Said Elon Musk on the plane back from Russia. 

> Pursuing Mars-related dreams, Elon Musk, Adeo Ressi and Jim Cantrell intended to make bussines with a russian company in order to adquire rockets. The original deal consisted of 3 ICBM for US$21M. But in the last meeting, russinans turned that into US$21M per ICBM, along with the taunt: "Oh, little boy, you don’t have the money?"

Reviewer: ```suggestion
Pursuing Mars-related dreams, Elon Musk, Adeo Ressi and Jim Cantrell intended to make business with a Russian company in order to acquire rockets. The original deal consisted of 3 ICBMs for US$21M. But in the last meeting, Russians turned that into US$21M per ICBM, along with the taunt: "Oh, little boy, you don’t have the money?"
```

> So... no exactly rockets, but similar words came out of my teammate's mouth.

Reviewer 2: `not exactly`

>"I think we can build an exception monitoring tool ourselves."

And what's more, why not offer it to our whole community? Let's make this Open Source. 

And that is how [ExceptionHunter](https://github.com/rootstrap/exception_hunter) was born. 

[![ExceptionHunter](https://github.com/rootstrap/exception_hunter/blob/develop/app/assets/images/exception_hunter/logo.png?raw=true)](https://github.com/rootstrap/exception_hunter)

### Installation
You can get it as a [gem](https://rubygems.org/gems/exception_hunter) for any of your ruby projects! It is as easy as adding to your gemfile:
```sh
gem 'exception_hunter'
```
and run:
```sh
$ rails generate exception_hunter:install
```

> And don't forget to include [devise](https://github.com/heartcombo/devise), in case you intend to use our build-in authentication. 

Reviewer 2: `built-in`

> Oh but if you do, don't worry, we will include it for you. 

Reviewer 2: I didn't understand this sentence. In the previous one, you say you should include devise if you want to use the built-in authentication, but then in this sentence you're saying you'll include it for the user? 🤔

Author: `if you do` refers to the action forgetting.
Like if you do forget, we will include it for you.
Any suggestions on clearer ways to specify this 😄 ?

Reviewer 2: Ahhh I understood `if you do include`, not `if you do forget`. Got it! Maybe just add the `forget`? `but if you do forget`?

> Happy tracking.

Reviewer: Maybe we can add a small section about our motivation while building this tool, we don't want to reinvent the wheel but rather help hobby projects and small MVP get exception tracking out of the box.

An increase in software applications needs is appreciable nowadays. People want software to make life easier, thus the user experience must be rated as FFF: friendly, fast, and fine. Easy right? No.
Every developer wishes their software to work like a charm, but since developers are human (at least for now), it is known for them to make mistakes. 

And mistakes, normally sabotage the user experience... But raise up, what doesn't kill you makes you stronger. 

The best way to go is by keeping these mistakes at bay. By monitoring them, awareness is key. But keeping record of hundreds maybe thousands (hopefully not) of errors, doesn't sound like an easy nor pleasant task. Yet diving through endless lines of logs isn't something you wish to someone, not even your worst enemy. There are easier ways, plenty of exception management tools are waiting for you to include them in your projects. 

Honeybadger, Sentry, Rollbar, are just a few names that resonate in this exception monitoring world. And don't take me wrong, they are great! But somehow all of them involve certain budget.  

 >“I think we can build a rocket ourselves.” 
 

> Said Elon Musk on the plane back from Russia. 

Reviewer 3: ```suggestion
Elon Musk said on the plane back from Russia. 
```

> Pursuing Mars-related dreams, Elon Musk, Adeo Ressi and Jim Cantrell intended to make business with a Russian company in order to acquire rockets. The original deal consisted of 3 ICBM for US$21M. But in the last meeting, Russians turned that into US$21M per ICBM, along with the taunt: "Oh, little boy, you don’t have the money?"

Reviewer 3: ```suggestion
Pursuing Mars-related dreams, Elon Musk, Adeo Ressi, and Jim Cantrell intended to make business with a Russian company in order to acquire rockets. The original deal consisted of 3 ICBMs for US$21M. But in the last meeting, Russians turned that into US$21M per ICBM, along with the taunt: "Oh, little boy, you don’t have the money?"
```

So... not exactly rockets, but similar words came out of my teammate's mouth.

>"I think we can build an exception monitoring tool ourselves."

And what's more, why not offer it to our whole community? Let's make this Open Source. 

> And that is how [ExceptionHunter](https://github.com/rootstrap/exception_hunter) was born. A simple dashboard, containing exceptions and errors occurred at your application grouped by similarity, along with tags differentiating among web-related, worker-related and manual. All distributed in a different tabs so the look up is easier. And you might also search through well-detailed independant paginated incidences from each error group.

Reviewer 3: ```suggestion
And that is how [ExceptionHunter](https://github.com/rootstrap/exception_hunter) was born. A simple dashboard, containing exceptions and errors occurred at your application grouped by similarity, along with tags differentiating among web-related, worker-related, and manual. All distributed in different tabs so the lookup is easier. And you might also search through well-detailed independent paginated incidences from each error group.
```

[![ExceptionHunter](https://github.com/rootstrap/exception_hunter/blob/develop/app/assets/images/exception_hunter/logo.png?raw=true)](https://github.com/rootstrap/exception_hunter)

### Installation
You can get it as a [gem](https://rubygems.org/gems/exception_hunter) for any of your ruby projects! It is as easy as adding to your gemfile:
```sh
gem 'exception_hunter'
```
and run:
```sh
$ rails generate exception_hunter:install
```

And don't forget to include [devise](https://github.com/heartcombo/devise), in case you intend to use our built-in authentication. 
Oh but if you do forget, don't worry, we will include it for you.

### Motivation

> So what exactly drove us to undertake such a proyect? We don't intend to reinvent the wheel, a wide variety of versions for this tool already exist, as said. Yet in our company, we can perceive a strong and growing open source culture that we identify with and support. 

Reviewer 3: ```suggestion
So what exactly drove us to undertake such a project? We don't intend to reinvent the wheel, a wide variety of versions for this tool already exist, as said. Yet in our company, we can perceive a strong and growing open-source culture that we identify with and support. 
```

> We hope that by sharing this project, we can give back to the community and help anyone developing hobby projects and small MVPs, where you may not want to pay for an external serivce, to get exception tracking right out of the box, without much effort.

Reviewer 3: ```suggestion
We hope that by sharing this project, we can give back to the community and help anyone developing hobby projects and small MVPs, where you may not want to pay for an external service, to get exception tracking right out of the box, without much effort.
```

> In case you are interested in the subject, stay tuned for incoming posts where we explain how we built ExceptionHunter from the ground up, along with some technical matters.

Reviewer 3: ```suggestion
If you are interested in the subject, stay tuned for incoming posts where we explain how we built ExceptionHunter from the ground up, along with some technical matters.
```
